Some fetch guide refactorings (#25819)
mdn/content: files/en-us/web/api/fetch_api/using_fetch/index.md

title: Using the Fetch API
slug: Web/API/Fetch_API/Using_Fetch
page-type: guide
---

{{DefaultAPISidebar("Fetch API")}}

The [Fetch API](/en-US/docs/Web/API/Fetch_API) provides a JavaScript interface for accessing and manipulating parts of the [protocol](/en-US/docs/Glossary/Protocol), such as requests and responses. It also provides a global {{domxref("fetch()")}} method that provides an easy, logical way to fetch resources asynchronously across the network.

> This kind of functionality was previously achieved using {{domxref("XMLHttpRequest")}}. Fetch provides a better alternative that can be easily used by other technologies such as [service workers](/en-US/docs/Web/API/Service_Worker_API). Fetch also provides a single logical place to define other HTTP-related concepts such as [CORS](/en-US/docs/Web/HTTP/CORS) and extensions to HTTP.

Reviewer: ```suggestion
Unlike {{domxref("XMLHttpRequest")}} that is a callback-based API, Fetch is promise-based and provides a better alternative that can be easily used in [service workers](/en-US/docs/Web/API/Service_Worker_API). Fetch also integrates advanced HTTP concepts such as [CORS](/en-US/docs/Web/HTTP/CORS) and other extensions to HTTP.
```

- `Content-Language`
- `Content-Type` with a value of `application/x-www-form-urlencoded`, `multipart/form-data`, or `text/plain`

## Aborting a fetch

> To abort incomplete `fetch()`, and even `XMLHttpRequest`, operations, use the {{DOMxRef("AbortController")}} and {{DOMxRef("AbortSignal")}} interfaces.

Reviewer: > and even `XMLHttpRequest`

Should we remove this phrase? Yes, we can pass an `AbortSignal` to an `XMLHttpRequest` but in a lot of cases calling `xhr.abort()` or `xhr.upload.abort()` is good enough. (And XHR uses the Fetch API behind the scene).

```suggestion
To abort incomplete `fetch()` operations, use the {{DOMxRef("AbortController")}} and {{DOMxRef("AbortSignal")}} interfaces.
```

```js
const controller = new AbortController();
const signal = controller.signal;
const url = "video.mp4";

> const downloadBtn = document.querySelector(".download");
> const abortBtn = document.querySelector(".abort");

Reviewer: Technically correct (so it is a nit), I would be surprised that `download` and `abort` would be classes. I would rather assume that they are IDs:
```suggestion
const downloadBtn = document.querySelector("#download");
const abortBtn = document.querySelector("#abort");
```

or using `getElementById()`.

Author: Sure, yes. I stole this bit from https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal#aborting_a_fetch_operation_using_an_explicit_signal :).

Use {{domxref("fetch()")}} to POST JSON-encoded data.

```js
async function postJSON(data) {
  try {
    const response = await fetch("https://example.com/profile", {
      method: "POST", // or 'PUT'
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(data),
    });

    const result = await response.json();
    console.log("Success:", result);
  } catch (error) {
    console.error("Error:", error);
  }
}

const data = { username: "example" };
postJSON(data);
```

## Uploading a file

Files can be uploaded using an HTML `<input type="file" />` input element, {{domxref("FormData.FormData","FormData()")}} and {{domxref("fetch()")}}.

```js
async function upload(formData) {
  try {
    const response = await fetch("https://example.com/profile/avatar", {
      method: "PUT",
      body: formData,
    });
    const result = await response.json();
    console.log("Success:", result);
  } catch (error) {
    console.error("Error:", error);
  }
}

const formData = new FormData();
const fileField = document.querySelector('input[type="file"]');

formData.append("username", "abc123");
formData.append("avatar", fileField.files[0]);

> upload(formData);

Reviewer: Is there a risk that this snippet doesn't work on the main thread anymore, while the original one did?

(Not an expert on this wrt to async function)

Author: Why? I think the async/await version is essentially the same, from this point of view, isn't it? (I mean, it doesn't actually block the thread.) Would love to know if I am wrong.

Reviewer: Because of this: https://v8.dev/features/top-level-await (that is not, yet, allowed). 

But I can be mistaken.

Author: But this isn't top-level await, is it?

> [Top-level await](https://github.com/tc39/proposal-top-level-await) enables developers to use the await keyword outside of async functions.

...I'm not using `await` outside an async function am I?

>   i++;

Reviewer: I'm not sure if it is better (and GitHub prevents me from making a proper suggestion here). What about:

```js
for (const [i, photo] of photos.files..entries()) {
  formData.append(`photos_${i}`, photo);
}
```

Author: Yes, that is better (less state).

Author: -> https://github.com/mdn/content/pull/25819/commits/2b4a26b3069992ee4209f451d911c44f9a6cd383

Reviewer: I have not tested it.

Author: Ah, a bit of testing suggests we need `Array.from(photos.files).entries`.

Author: -> https://github.com/mdn/content/pull/25819/commits/55f1950fd27fa600beb30aef82b45e993d2a17f3

Reviewer: Yes, collections from forms may be "incomplete". It is on my plans for Q2/Q3 to sort this out.

}
```

## Differences from `jQuery.ajax()`

The `fetch` specification differs from `jQuery.ajax()` in the following significant ways:

> - The Promise returned from `fetch()` **won't reject on HTTP error status** even if the response is an HTTP 404 or 500. Instead, as soon as the server responds with headers, the Promise will resolve normally (with the {{domxref("Response/ok", "ok")}} property of the response set to false if the response isn't in the range 200–299), and it will only reject on network failure or if anything prevented the request from completing.

Reviewer: NIT. I'm not a fan of bold for emphasis. HTML's semantic `<em>` element is usually marked with italics.
```suggestion
- The Promise returned from `fetch()` _won't reject on HTTP error status_ even if the response is an HTTP 404 or 500. Instead, as soon as the server responds with headers, the Promise will normally resolve (with the {{domxref("Response/ok", "ok")}} property of the response set to false if the response isn't in the range 200–299), and it will only reject on network failure or if anything prevented the request from completing.
```

Author: "normally resolve" is wrong, since it implies "usually" while of course the promise always resolves. Just "resolve" is better than the original though.

And I don't like any emphasis here though I'm a bit of a typographical minimalist :).

Reviewer: I'm fine with a single resolve; and also with no emphasis (my only strong feeling is about bold for emphasis)

> - The Promise returned from `fetch()` **won't reject on HTTP error status** even if the response is an HTTP 404 or 500. Instead, as soon as the server responds with headers, the Promise will resolve normally (with the {{domxref("Response/ok", "ok")}} property of the response set to false if the response isn't in the range 200–299), and it will only reject on network failure or if anything prevented the request from completing.

Author: ```suggestion
- The promise returned from `fetch()` won't reject on HTTP errors even if the response is an HTTP 404 or 500. Instead, as soon as the server responds with headers, the promise will resolve (with the {{domxref("Response/ok", "ok")}} property of the response set to `false` if the response isn't in the range 200–299). The promise will only reject on network failure or if anything prevented the request from completing.
```

honestly once we start getting into the prose I want to rewrite all of it.
